Document Deployment Process (#1953)
CDCgov/prime-reportstream: DEPLOYMENTS.md

## When do ReportStream deployments take place?

ReportStream production deployments take place on Tuesdays and Thursday around 10am EST.

> The lower environments, test and staging, are automatically deployed on every merge to `master` (a.k.a. continuous deployment).

Reviewer: prefer:
```suggestion
The testing and staging environments are automatically deployed on every merge into `master` through our Continuous Deployment pipeline.
```

## How is ReportStream deployed?

> ReportStream is deployed via [this GitHub Action](.github/workflows/release.yml).

Reviewer: Consider:
```Markdown
We automatically deploy changes through [GitHub Actions](.github/workflows/release.yml) based on changes in target branches as described in the table below. These changes can only enter these branches through successfully reviewed Pull Requests.

| Changes are merged into branch | Changes get deployed into environment(s) |
|:--|:--|
| `master` | test and staging |
| `production` | production |
```

> ReportStream is deployed via [this GitHub Action](.github/workflows/release.yml).
> 
> The GitHub Action automatically deploys changes in the following branches to the specified environments:
> 
> | Branch | Environment |
> |:--|:--|
> | `master` | test, staging |
> | `production` | production |
> 
> Code merged to the above branches always go through a PR review.

Author: ```suggestion
We automatically deploy changes through [GitHub Actions](.github/workflows/release.yml) based on changes in target branches as described in the table below. These changes can only enter these branches through successfully reviewed Pull Requests.

| Changes are merged into branch | Changes get deployed into environment(s) |
|:--|:--|
| `master` | test and staging |
| `production` | production |
```

The deployment process is zero-downtime with a blue/green deployment process used [via Azure Function deployment slots](https://docs.microsoft.com/en-us/azure/azure-functions/functions-deployment-slots). The deployment process is [documented in PR #1318](https://github.com/CDCgov/prime-reportstream/issues/1318).

> Smoke tests are run manually after a deployment, but in the near future the will run automatically before directing traffic to the newly deployed code ([work tracked in #656](https://github.com/CDCgov/prime-reportstream/issues/656)).

Reviewer: NOTE: you can just do #<issuenumber> and GitHub will render it as a link for you...

```suggestion
Our CI/CD pipeline does not run the Smoke Tests; therefor you *must* run the Smoke tests manually after every deployment. While this will change in the near future (i.e. in the future, they _will_ be included in the pipeline), until that time, *you must run the Smoke tests manually and validate they complete successfully* before directing _any_ traffic to the newly deployed code.
You can track #656 for the status on running the Smoke tests automatically.
```

Author: That's what I initially tried, and maybe it doesn't work for PRs, but the issues were not linked.

Reviewer: ah, okido, didn't know that... TIL

## What is deployed?

The GitHub Action deploys changes in the following directories:

| Directory | Deployment Process |
|:--|:--|
| [/prime-router](/prime-router) | New version released to Azure Functions |
| [/frontend](/frontend) | Static site served via an Azure Storage Account |

> | [/operations](/operations) | Not deployed. [Coming soon in #578](https://github.com/CDCgov/prime-reportstream/issues/578) |

Reviewer: ```suggestion
| [/operations](/operations) | Not deployed. (Coming soon in, see #578) |
```

## What is our deployment process for production?

In preparation for our Tuesday and Thursday deployments, a cutoff time has been established.

| Deployment Window | Cutoff Time |
|:--|:--|
| Tuesday, 10am EST | Monday, 12pm EST |
| Thursday, 10am EST | Wednesday, 12pm EST |

> The cutoff time is automatically enforced via [this GitHub Action](.github/workflows/prepare_deployment_branch.yaml).

Reviewer: ```suggestion
The cutoff time is automatically enforced via automatic branching from `master` into a dedicated deployment branch targeting the `production` branch through a [GitHub Action](.github/workflows/prepare_deployment_branch.yaml).
```

> 1. At the specified cutoff a time, a PR is created to merge the contents of `master` to the `production` branch

Reviewer: Other suggestion: prefer (in Markdown) to name all numbered list items with `1.` instead of 1, 2, 3, ... Markdown takes care of properly numbering things. It makes it so that if you insert things in the future (or remove them), you don't have to touch unrelated lines.

```suggestion
1. At the specified cut-off time (Mondays and Wednesdays at noon ET), the GitHub action creates a new branch named `deployment/YYYY-MM-DD` (where the YYYY-MM-DD is `today + 1day`, i.e. the date of the deployment, not the date of the branching) which branches from `master`. This branch now contains everything that was present in `master` at that cut-off time. This is the content that is/will be part of the production deployment.
1. A new PR from the deployment branch is filed to merge `deployment/YYYY-MM-DD` into `production`. The PR has title `"Deployment of YYYY-MM-DD"` and is tagged with the [`deployment` tag](https://github.com/CDCgov/prime-reportstream/issues?q=label%3Adeployment).
```

Reviewer 2: Thanks - it helps to explain that this is both a PR and a new branch.

2. The contents of `master` is deployed to the staging environment for verification
    * Manual testing takes place
3. The PR is reviewed by the team

> 4. The PR is merged during the specified deployment window by an approved team member

Reviewer 2: ```suggestion
4. The PR is merged during the specified deployment window by an approved team member.   Note: Special care must be taken if the release includes schema changes that add or remove data elements for receivers whose data is waiting for the `batch` step at the time of deployment.    Inconsistencies between data models and actual internal data can cause errors.  At this time, our solution has been to do the deployments at times when there is no work-in-progress (WIP) in the system.
```
